src/PostgREST/Query/QueryBuilder.hs

```diff
@@ -81,12 +82,25 @@ getSelectsJoins rr@(Node ReadPlan{relName, relToParent=Just rel, relAggAlias, re
   in
   (sel:selects, joi:joins)
 
+pgSelectFromJson :: S.Set TypedField -> SQL.Snippet
+pgSelectFromJson fields =
+  SQL.sql "SELECT " <> parsedCols <> " " <>
+  (if S.null fields
+    -- When we are inserting no columns (e.g. using default values), we can't use our ordinary `json_to_recordset`
+    -- because it can't extract records with no columns (there's no valid syntax for the `AS (colName colType,...)`
+    -- part). But we still need to ensure as many rows are created as there are array elements.
+    then SQL.sql ("FROM json_array_elements (" <> selectBody <> ") _ ")
+    else SQL.sql ("FROM json_to_recordset (" <> selectBody <> ") AS _ " <> "(" <> typedCols <> ") ")
```

Comment:
This reminds me of https://github.com/PostgREST/postgrest/issues/1567#issuecomment-739478169, not having defaults is still a problem.

Would `json_to_recordset` help or make it harder to do the above suggestion later on?